File: slp/backend-nestjs/src/helpers/question-validation.helper.ts
import { BadRequestException } from '@nestjs/common';

type SupportedType = 'multiple_choice' | 'single_choice' | 'true_false' | 'fill_blank' | 'ordering' | 'matching' | 'flashcard';

interface Option {
  id: string;
  text: string;
}

interface MatchingPair {
  id: number;
  left: string;
  right: string;
}

interface OrderingItem {
  order_id: number;
  text: string;
}

interface Flashcard {
  front: string;
  back: string;
}

interface Metadata {
  options?: Option[];
  correctAnswers?: string[];
  correctAnswer?: boolean;
  keywords?: string[];
  answers?: string[];
  items?: OrderingItem[];
  pairs?: MatchingPair[];
  front?: string;
  back?: string;
}

export class QuestionValidationHelper {
  private static supportedTypes: SupportedType[] = [
    'multiple_choice',
    'single_choice',
    'true_false',
    'fill_blank',
    'ordering',
    'matching',
    'flashcard',
  ];

  static validateSnapshot(snapshot: any): void {
    if (!snapshot || typeof snapshot !== 'object') {
      throw new BadRequestException('Question snapshot must be a valid object');
    }

    const { type, content, metadata } = snapshot;

    // Validate type
    if (!type || typeof type !== 'string') {
      throw new BadRequestException('Question snapshot must contain a string "type" field.');
    }
    if (!this.supportedTypes.includes(type as SupportedType)) {
      throw new BadRequestException(`Unsupported question type: '${type}'. Supported types: ${this.supportedTypes.join(', ')}.`);
    }

    // Validate content
    if (!content || typeof content !== 'string' || content.trim() === '') {
      throw new BadRequestException('Question snapshot must contain a non-empty string "content" field.');
    }

    // Validate metadata
    if (!metadata || typeof metadata !== 'object') {
      throw new BadRequestException('Question snapshot must contain a "metadata" object.');
    }

    // Type-specific validation
    switch (type) {
      case 'multiple_choice':
        this.validateMultipleChoice(metadata);
        break;
      case 'single_choice':
        this.validateSingleChoice(metadata);
        break;
      case 'true_false':
        this.validateTrueFalse(metadata);
        break;
      case 'fill_blank':
        this.validateFillBlank(metadata, content);
        break;
      case 'ordering':
        this.validateOrdering(metadata);
        break;
      case 'matching':
        this.validateMatching(metadata);
        break;
      case 'flashcard':
        this.validateFlashcard(metadata);
        break;
    }
  }

  private static validateMultipleChoice(metadata: Metadata): void {
    const options = metadata.options;
    if (!options || !Array.isArray(options) || options.length === 0) {
      throw new BadRequestException('Multiple choice question must have a non-empty "options" array.');
    }
    const optionIds = new Set<string>();
    for (let i = 0; i < options.length; i++) {
      const opt = options[i];
      if (!opt.id || typeof opt.id !== 'string' || opt.id.trim() === '') {
        throw new BadRequestException(`Multiple choice options[${i}] must have a non-empty string "id".`);
      }
      if (optionIds.has(opt.id)) {
        throw new BadRequestException(`Multiple choice options contain duplicate id: "${opt.id}".`);
      }
      optionIds.add(opt.id);
      if (!opt.text || typeof opt.text !== 'string' || opt.text.trim() === '') {
        throw new BadRequestException(`Multiple choice options[${i}] must have a non-empty string "text".`);
      }
    }

    const correctAnswers = metadata.correctAnswers;
    if (!correctAnswers || !Array.isArray(correctAnswers) || correctAnswers.length === 0) {
      throw new BadRequestException('Multiple choice question must have a non-empty "correctAnswers" array.');
    }
    for (let i = 0; i < correctAnswers.length; i++) {
      const ans = correctAnswers[i];
      if (typeof ans !== 'string' || ans.trim() === '') {
        throw new BadRequestException(`Multiple choice correctAnswers[${i}] must be a non-empty string.`);
      }
      if (!optionIds.has(ans)) {
        throw new BadRequestException(`Multiple choice correctAnswers references id "${ans}" which does not exist in options.`);
      }
    }
  }

  private static validateSingleChoice(metadata: Metadata): void {
    const options = metadata.options;
    if (!options || !Array.isArray(options) || options.length === 0) {
      throw new BadRequestException('Single choice question must have a non-empty "options" array.');
    }
    const optionIds = new Set<string>();
    for (let i = 0; i < options.length; i++) {
      const opt = options[i];
      if (!opt.id || typeof opt.id !== 'string' || opt.id.trim() === '') {
        throw new BadRequestException(`Single choice options[${i}] must have a non-empty string "id".`);
      }
      if (optionIds.has(opt.id)) {
        throw new BadRequestException(`Single choice options contain duplicate id: "${opt.id}".`);
      }
      optionIds.add(opt.id);
      if (!opt.text || typeof opt.text !== 'string' || opt.text.trim() === '') {
        throw new BadRequestException(`Single choice options[${i}] must have a non-empty string "text".`);
      }
    }

    const correctAnswers = metadata.correctAnswers;
    if (!correctAnswers || !Array.isArray(correctAnswers) || correctAnswers.length !== 1) {
      throw new BadRequestException('Single choice question must have exactly one "correctAnswers" entry.');
    }
    const ans = correctAnswers[0];
    if (typeof ans !== 'string' || ans.trim() === '') {
      throw new BadRequestException('Single choice correctAnswers[0] must be a non-empty string.');
    }
    if (!optionIds.has(ans)) {
      throw new BadRequestException(`Single choice correctAnswers references id "${ans}" which does not exist in options.`);
    }
  }

  private static validateTrueFalse(metadata: Metadata): void {
    if (typeof metadata.correctAnswer !== 'boolean') {
      throw new BadRequestException('True/false question must have a boolean "correctAnswer" field.');
    }
  }

  private static validateFillBlank(metadata: Metadata, content: string): void {
    const keywords = metadata.keywords;
    if (!keywords || !Array.isArray(keywords) || keywords.length === 0) {
      throw new BadRequestException('Fill blank question must have a non-empty "keywords" array.');
    }
    for (let i = 0; i < keywords.length; i++) {
      const kw = keywords[i];
      if (typeof kw !== 'string' || kw.trim() === '') {
        throw new BadRequestException(`Fill blank keywords[${i}] must be a non-empty string.`);
      }
      if (!content.toLowerCase().includes(kw.toLowerCase())) {
        throw new BadRequestException(`Fill blank keyword "${kw}" not found in content.`);
      }
    }
    // answers are optional but if provided must be non-empty strings
    const answers = metadata.answers;
    if (answers !== undefined) {
      if (!Array.isArray(answers) || answers.length === 0) {
        throw new BadRequestException('Fill blank "answers" if provided must be a non-empty array.');
      }
      for (let i = 0; i < answers.length; i++) {
        if (typeof answers[i] !== 'string' || answers[i].trim() === '') {
          throw new BadRequestException(`Fill blank answers[${i}] must be a non-empty string.`);
        }
      }
    }
  }

  private static validateOrdering(metadata: Metadata): void {
    const items = metadata.items;
    if (!items || !Array.isArray(items) || items.length === 0) {
      throw new BadRequestException('Ordering question must have a non-empty "items" array.');
    }
    const orderIds = new Set<number>();
    for (let i = 0; i < items.length; i++) {
      const item = items[i];
      if (typeof item !== 'object' || item === null) {
        throw new BadRequestException(`Ordering items[${i}] must be an object.`);
      }
      if (typeof item.order_id !== 'number' || item.order_id < 1) {
        throw new BadRequestException(`Ordering items[${i}] must have an integer "order_id" >= 1.`);
      }
      if (orderIds.has(item.order_id)) {
        throw new BadRequestException(`Ordering items contains duplicate order_id: ${item.order_id}.`);
      }
      orderIds.add(item.order_id);
      if (typeof item.text !== 'string' || item.text.trim() === '') {
        throw new BadRequestException(`Ordering items[${i}] must have a non-empty string "text".`);
      }
    }
  }

  private static validateMatching(metadata: Metadata): void {
    const pairs = metadata.pairs;
    if (!pairs || !Array.isArray(pairs) || pairs.length === 0) {
      throw new BadRequestException('Matching question must have a non-empty "pairs" array.');
    }
    const ids = new Set<number>();
    for (let i = 0; i < pairs.length; i++) {
      const pair = pairs[i];
      if (typeof pair !== 'object' || pair === null) {
        throw new BadRequestException(`Matching pairs[${i}] must be an object.`);
      }
      if (typeof pair.id !== 'number' || pair.id < 1) {
        throw new BadRequestException(`Matching pairs[${i}] must have an integer "id" >= 1.`);
      }
      if (ids.has(pair.id)) {
        throw new BadRequestException(`Matching pairs contains duplicate id: ${pair.id}.`);
      }
      ids.add(pair.id);
      if (typeof pair.left !== 'string' || pair.left.trim() === '') {
        throw new BadRequestException(`Matching pairs[${i}] must have a non-empty string "left".`);
      }
      if (typeof pair.right !== 'string' || pair.right.trim() === '') {
        throw new BadRequestException(`Matching pairs[${i}] must have a non-empty string "right".`);
      }
    }
  }

  private static validateFlashcard(metadata: Metadata): void {
    const front = metadata.front;
    const back = metadata.back;
    if (typeof front !== 'string' || front.trim() === '') {
      throw new BadRequestException('Flashcard must have a non-empty string "front" in metadata.');
    }
    if (typeof back !== 'string' || back.trim() === '') {
      throw new BadRequestException('Flashcard must have a non-empty string "back" in metadata.');
    }
  }
}

File: slp/backend-dotnet/Features/Question/QuestionValidationHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace backend_dotnet.Features.Question;

/// <summary>
/// Validates a question snapshot against the canonical schema.
/// No legacy field aliases are accepted — all field names are strict.
/// See QUESTION_SCHEMA.md for the full specification.
/// </summary>
public static class QuestionValidationHelper
{
    private static readonly HashSet<string> SupportedTypes = new()
    {
        "multiple_choice",
        "single_choice",
        "true_false",
        "fill_blank",
        "ordering",
        "matching",
        "flashcard",
    };

    /// <summary>
    /// Validates the full snapshot JSON string.
    /// Throws <see cref="ArgumentException"/> with a descriptive message on any violation.
    /// </summary>
    public static void ValidateSnapshot(string snapshotJson)
    {
        JsonElement root;
        try
        {
            root = JsonDocument.Parse(snapshotJson).RootElement;
        }
        catch (JsonException ex)
        {
            throw new ArgumentException("Question snapshot is not valid JSON.", ex);
        }

        // ── type ────────────────────────────────────────────────────────────
        if (!root.TryGetProperty("type", out var typeEl) || typeEl.ValueKind != JsonValueKind.String)
            throw new ArgumentException("Question snapshot must contain a string 'type' field.");

        var type = typeEl.GetString()!;
        if (!SupportedTypes.Contains(type))
            throw new ArgumentException(
                $"Unsupported question type: '{type}'. " +
                $"Supported types: {string.Join(", ", SupportedTypes)}.");

        // ── content ─────────────────────────────────────────────────────────
        if (!root.TryGetProperty("content", out var contentEl) || contentEl.ValueKind != JsonValueKind.String)
            throw new ArgumentException("Question snapshot must contain a string 'content' field.");

        var content = contentEl.GetString()!;
        if (string.IsNullOrWhiteSpace(content))
            throw new ArgumentException("Question snapshot 'content' must not be empty.");

        // ── metadata ────────────────────────────────────────────────────────
        if (!root.TryGetProperty("metadata", out var metaEl) || metaEl.ValueKind != JsonValueKind.Object)
            throw new ArgumentException("Question snapshot must contain a 'metadata' object.");

        // ── type-specific validation ─────────────────────────────────────────
        switch (type)
        {
            case "multiple_choice": ValidateMultipleChoice(metaEl); break;
            case "single_choice": ValidateSingleChoice(metaEl); break;
            case "true_false": ValidateTrueFalse(metaEl); break;
            case "fill_blank": ValidateFillBlank(metaEl, content); break;
            case "ordering": ValidateOrdering(metaEl); break;
            case "matching": ValidateMatching(metaEl); break;
            case "flashcard": ValidateFlashcard(metaEl); break;
        }
    }

    // Kept for backward-compat call sites that pass type+content+metadataJson separately.
    public static void ValidateQuestionMetadata(string type, string content, string? metadataJson)
    {
        // Re-assemble a minimal snapshot and delegate to the unified validator.
        var assembled = metadataJson == null || metadataJson == "null"
            ? $@"{{""type"":""{type}"",""content"":""{EscapeJson(content)}"",""metadata"":""{{}}""}}"
            : $@"{{""type"":""{type}"",""content"":""{EscapeJson(content)}"",""metadata"":{metadataJson}}}";

        ValidateSnapshot(assembled);
    }

    // ── Validators ────────────────────────────────────────────────────────────

    private static void ValidateMultipleChoice(JsonElement meta)
    {
        var options = RequireNonEmptyArray(meta, "options", "Multiple choice");
        var optionIds = ValidateOptions(options, "Multiple choice");

        var correctAnswers = RequireNonEmptyArray(meta, "correctAnswers", "Multiple choice");
        ValidateAnswerIdsExist(correctAnswers, optionIds, "correctAnswers", "Multiple choice");
    }

    private static void ValidateSingleChoice(JsonElement meta)
    {
        var options = RequireNonEmptyArray(meta, "options", "Single choice");
        var optionIds = ValidateOptions(options, "Single choice");

        var correctAnswers = RequireNonEmptyArray(meta, "correctAnswers", "Single choice");
        if (correctAnswers.Count != 1)
            throw new ArgumentException("Single choice question 'correctAnswers' must contain exactly one id.");

        ValidateAnswerIdsExist(correctAnswers, optionIds, "correctAnswers", "Single choice");
    }

    private static void ValidateTrueFalse(JsonElement meta)
    {
        if (!meta.TryGetProperty("correctAnswer", out var el) ||
            (el.ValueKind != JsonValueKind.True && el.ValueKind != JsonValueKind.False))
            throw new ArgumentException(
                "True/false question must have a boolean 'correctAnswer' field.");
    }

    private static void ValidateFillBlank(JsonElement meta, string content)
    {
        // keywords — words to blank out in display; must appear in content
        var keywords = RequireNonEmptyStringArray(meta, "keywords", "Fill blank");
        foreach (var kw in keywords)
        {
            if (!content.Contains(kw, StringComparison.OrdinalIgnoreCase))
                throw new ArgumentException(
                    $"Fill blank 'keywords' entry \"{kw}\" was not found in the question content.");
        }

        // answers — grading values; must be a non-empty string array
        //RequireNonEmptyStringArray(meta, "answers", "Fill blank");
    }

    private static void ValidateOrdering(JsonElement meta)
    {
        if (!meta.TryGetProperty("items", out var itemsEl) || itemsEl.ValueKind != JsonValueKind.Array)
            throw new ArgumentException("Ordering question must have an 'items' array.");

        var items = itemsEl.EnumerateArray().ToList();
        if (items.Count == 0)
            throw new ArgumentException("Ordering question 'items' must not be empty.");

        var orderIds = new HashSet<int>();
        foreach (var (item, idx) in items.Select((x, i) => (x, i)))
        {
            if (item.ValueKind != JsonValueKind.Object)
                throw new ArgumentException($"Ordering 'items[{idx}]' must be an object.");

            if (!item.TryGetProperty("order_id", out var orderIdEl) ||
                orderIdEl.ValueKind != JsonValueKind.Number ||
                !orderIdEl.TryGetInt32(out var orderId) || orderId < 1)
                throw new ArgumentException(
                    $"Ordering 'items[{idx}]' must have an integer 'order_id' ≥ 1.");

            if (!orderIds.Add(orderId))
                throw new ArgumentException(
                    $"Ordering 'items' contains duplicate order_id: {orderId}.");

            if (!item.TryGetProperty("text", out var textEl) || textEl.ValueKind != JsonValueKind.String ||
                string.IsNullOrWhiteSpace(textEl.GetString()))
                throw new ArgumentException($"Ordering 'items[{idx}]' must have a non-empty string 'text'.");
        }
    }

    private static void ValidateMatching(JsonElement meta)
    {
        if (!meta.TryGetProperty("pairs", out var pairsEl) || pairsEl.ValueKind != JsonValueKind.Array)
            throw new ArgumentException("Matching question must have a 'pairs' array.");

        var pairs = pairsEl.EnumerateArray().ToList();
        if (pairs.Count == 0)
            throw new ArgumentException("Matching question 'pairs' must not be empty.");

        var ids = new HashSet<int>();
        foreach (var (pair, idx) in pairs.Select((x, i) => (x, i)))
        {
            if (pair.ValueKind != JsonValueKind.Object)
                throw new ArgumentException($"Matching 'pairs[{idx}]' must be an object.");

            if (!pair.TryGetProperty("id", out var idEl) ||
                idEl.ValueKind != JsonValueKind.Number ||
                !idEl.TryGetInt32(out var id) || id < 1)
                throw new ArgumentException(
                    $"Matching 'pairs[{idx}]' must have an integer 'id' ≥ 1.");

            if (!ids.Add(id))
                throw new ArgumentException(
                    $"Matching 'pairs' contains duplicate id: {id}.");

            foreach (var field in new[] { "left", "right" })
            {
                if (!pair.TryGetProperty(field, out var fieldEl) ||
                    fieldEl.ValueKind != JsonValueKind.String ||
                    string.IsNullOrWhiteSpace(fieldEl.GetString()))
                    throw new ArgumentException(
                        $"Matching 'pairs[{idx}]' must have a non-empty string '{field}'.");
            }
        }
    }

    private static void ValidateFlashcard(JsonElement meta)
    {
        foreach (var field in new[] { "front", "back" })
        {
            if (!meta.TryGetProperty(field, out var el) ||
                el.ValueKind != JsonValueKind.String ||
                string.IsNullOrWhiteSpace(el.GetString()))
                throw new ArgumentException(
                    $"Flashcard must have a non-empty string '{field}' in metadata.");
        }
    }

    // ── Shared helpers ────────────────────────────────────────────────────────

    private static List<JsonElement> RequireNonEmptyArray(JsonElement meta, string field, string label)
    {
        if (!meta.TryGetProperty(field, out var el) || el.ValueKind != JsonValueKind.Array)
            throw new ArgumentException($"{label} question must have a '{field}' array.");

        var list = el.EnumerateArray().ToList();
        if (list.Count == 0)
            throw new ArgumentException($"{label} question '{field}' must not be empty.");

        return list;
    }

    /// <summary>Validates options shape and returns the set of option ids.</summary>
    private static HashSet<string> ValidateOptions(List<JsonElement> options, string label)
    {
        var ids = new HashSet<string>();
        foreach (var (opt, idx) in options.Select((x, i) => (x, i)))
        {
            if (opt.ValueKind != JsonValueKind.Object)
                throw new ArgumentException($"{label} 'options[{idx}]' must be an object.");

            if (!opt.TryGetProperty("id", out var idEl) || idEl.ValueKind != JsonValueKind.String ||
                string.IsNullOrWhiteSpace(idEl.GetString()))
                throw new ArgumentException(
                    $"{label} 'options[{idx}]' must have a non-empty string 'id' field.");

            var id = idEl.GetString()!;
            if (!ids.Add(id))
                throw new ArgumentException($"{label} 'options' contains duplicate id: \"{id}\".");

            if (!opt.TryGetProperty("text", out var textEl) || textEl.ValueKind != JsonValueKind.String ||
                string.IsNullOrWhiteSpace(textEl.GetString()))
                throw new ArgumentException(
                    $"{label} 'options[{idx}]' must have a non-empty string 'text' field.");
        }
        return ids;
    }

    private static void ValidateAnswerIdsExist(
        List<JsonElement> correctAnswers,
        HashSet<string> optionIds,
        string field,
        string label)
    {
        foreach (var (el, idx) in correctAnswers.Select((x, i) => (x, i)))
        {
            if (el.ValueKind != JsonValueKind.String || string.IsNullOrWhiteSpace(el.GetString()))
                throw new ArgumentException(
                    $"{label} '{field}[{idx}]' must be a non-empty string.");

            var id = el.GetString()!;
            if (!optionIds.Contains(id))
                throw new ArgumentException(
                    $"{label} '{field}' references id \"{id}\" which does not exist in 'options'.");
        }
    }

    private static List<string> RequireNonEmptyStringArray(JsonElement meta, string field, string label)
    {
        if (!meta.TryGetProperty(field, out var el) || el.ValueKind != JsonValueKind.Array)
            throw new ArgumentException($"{label} question must have a '{field}' array.");

        var list = el.EnumerateArray().ToList();
        if (list.Count == 0)
            throw new ArgumentException($"{label} question '{field}' must not be empty.");

        var result = new List<string>();
        foreach (var (item, idx) in list.Select((x, i) => (x, i)))
        {
            if (item.ValueKind != JsonValueKind.String || string.IsNullOrWhiteSpace(item.GetString()))
                throw new ArgumentException(
                    $"{label} '{field}[{idx}]' must be a non-empty string.");
            result.Add(item.GetString()!);
        }
        return result;
    }

    private static string EscapeJson(string s) =>
        s.Replace("\\", "\\\\").Replace("\"", "\\\"");
}